File: src/main/java/cz/vhromada/catalog/web/validator/constraints/DateRange.java
package cz.vhromada.catalog.web.validator.constraints;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import javax.validation.Constraint;
import javax.validation.Payload;

import cz.vhromada.catalog.web.validator.DateRangeValidator;

/**
 * An annotation represents date range constraint.
 *
 * @author Vladimir Hromada
 */
@Target(ElementType.FIELD)
@Retention(RetentionPolicy.RUNTIME)
@Constraint(validatedBy = DateRangeValidator.class)
@Documented
public @interface DateRange {

    /**
     * Returns error message template.
     *
     * @return error message template
     */
    String message() default "{cz.vhromada.catalog.validator.DateRange.message}";

    /**
     * Returns groups constraint belongs to.
     *
     * @return groups constraint belongs to
     */
    Class<?>[] groups() default {};

    /**
     * Returns payload associated to constraint.
     *
     * @return payload associated to constraint
     */
    Class<? extends Payload>[] payload() default {};

    /**
     * Returns valid minimal year.
     *
     * @return valid minimal year
     */
    int value();

}

File: src/main/kotlin/cz/vhromada/catalog/web/validator/constraints/DateRange.kt
package cz.vhromada.catalog.web.validator.constraints

import cz.vhromada.catalog.web.validator.DateRangeValidator
import javax.validation.Constraint
import javax.validation.Payload
import kotlin.reflect.KClass

/**
 * An annotation represents date range constraint.
 *
 * @author Vladimir Hromada
 */
@Target(AnnotationTarget.FIELD)
@Retention(AnnotationRetention.RUNTIME)
@Constraint(validatedBy = [DateRangeValidator::class])
@MustBeDocumented
annotation class DateRange(

        /**
         * Returns valid minimal year.
         *
         * @return valid minimal year
         */
        val value: Int,

        /**
         * Returns error message template.
         *
         * @return error message template
         */
        val message: String = "{cz.vhromada.catalog.validator.DateRange.message}",

        /**
         * Returns groups constraint belongs to.
         *
         * @return groups constraint belongs to
         */
        val groups: Array<KClass<*>> = [],

        /**
         * Returns payload associated to constraint.
         *
         * @return payload associated to constraint
         */
        @Suppress("unused")
        val payload: Array<KClass<out Payload>> = [])
